### PR TITLE
Feat/nx workspace

### PR DESCRIPTION
**Pull request by edbzn**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://www.conventionalcommits.org/en/v1.0.0/
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Chore
- [ ] Refactoring (no functional changes, no api changes)
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

No support for `workspace.json` definition file when using in Nx devtools.

Issue Number: N/A

## What is the new behavior?

Support `workspace.json` definition file.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

**Comment by dianjuar** [comment]
This is awesome!!!

@beeman @johanneshoppe You have to see this!!! 👀

File: src/engine/engine.ts [hunk 1]
```diff
@@ -85,10 +85,15 @@ function prepareOptions(origOptions: Schema, logger: logging.LoggerApi) {
     logger.info('Dry-run: No changes are applied at all.');
   }
 
-  return options;
+  return options as Schema;
 }
 
-function getOptionsString(options: Schema) {
+function getOptionsString(options: {
```
Comment (dianjuar) [inline src/engine/engine.ts:91]: Really good catch 🧠 

You gave an idea to create an Interface only with the params used by `npm publish`, which would be something like this `Pick<Schema, 'access' | 'tag' | 'otp' | 'dryRun'>`, this should be applied to the function `extractOnlyNPMOptions` in this same file

File: src/engine/engine.ts [hunk 2]
```diff
@@ -75,7 +75,7 @@ function extractOnlyNPMOptions({ access, tag, otp, dryRun }: Schema) {
   };
 }
 
-function prepareOptions(origOptions: Schema, logger: logging.LoggerApi) {
+function prepareOptions(origOptions: Schema, logger: logging.LoggerApi): Schema {
```
Comment (dianjuar) [inline src/engine/engine.ts:78]: Thanks the implicit returning type of typescript is already returning the `Schema` type but I think is better to put it explicitly like you did, thanks for making this 🙌

File: src/engine/engine.spec.ts [hunk 3]
```diff
@@ -2,7 +2,6 @@ import { logging } from '../__mocks__/utils/dev-kit-logger';
```
Comment (dianjuar) [inline src/engine/engine.spec.ts:2]: I rather have a specific type than `any` as a general rule when I'm coding In Typescript. I want to know the reasons that lead you to use `any`, maybe I am not seeing something.

File: src/deploy/actions.spec.ts [hunk 4]
```diff
@@ -50,7 +50,7 @@ describe('Deploy Angular apps', () => {
     });
```
Comment (dianjuar) [inline src/deploy/actions.spec.ts:50]: I rather have a specific type than `any` as a general rule when I'm coding In Typescript. I want to know the reasons that lead you to use `any`, maybe I am not seeing something.

File: src/deploy/schema.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+export interface Schema extends JsonObject {
+  configuration: string | null;
+  noBuild: boolean;
+  packageVersion: string | null;
+  tag: string | null;
+  access: string;
+  otp: string | null;
+  dryRun: boolean;
+}
```
[the remark this answers is not in the document]
Comment (edbzn) [reply src/deploy/schema.ts:11]: That's what I missed, thank you.

File: src/engine/engine.spec.ts (hunk 3 above)
Comment (edbzn) [reply src/engine/engine.spec.ts:2]: In test files I often use `any` when I want to create partial options just containing what's needed for the test case, it's easier to test some scenarios with the minimum boilerplate, but if you think it's better with a type I can rewrite it as `const baseSchema: Schema = { ... }`.

File: src/engine/engine.ts (hunk 1 above)
Comment (edbzn) [reply src/engine/engine.ts:91]: Nice suggestion, will do it.

File: src/engine/engine.spec.ts (hunk 3 above)
Comment (edbzn) [reply src/engine/engine.spec.ts:2]: What do you think?

File: src/deploy/actions.spec.ts (hunk 4 above)
Comment (edbzn) [reply src/deploy/actions.spec.ts:50]: Same as above, I can rewrite it if you think it's better with types.

**Comment by edbzn** [comment]
@dianjuar I applied all your suggested changes, that's better as it! :blush:

**Review by dianjuar — COMMENTED** [review]
@edbzn I think you accidentally remove the `workspace.json` file for the PR.

**Comment by edbzn** [comment]
@dianjuar Mmh no, I just added the ability to read the `workspace.json` file to get the workspace definition but I didn't added this file.

**Comment by edbzn** [comment]
I noticed that I missed some things to make it work for all Nx projects *including non-Angular libraries*. Actually the published directory is grabbed from the `dest` property in the project's `ng-package.json` file, we should fallback to the `outputPath` property in `workspace.json` definition file. 

Also we need to ensure that this builder is idempotent to suit our needs: automated release cycle for a monorepo. We will iterate and create other PRs, but because we want this builder for our [@jscutlery/semver](https://github.com/jscutlery/semver) plugin, I will re-fork this project under our organization and point to our git while these PRs are pending review, it will be easier for us. 

However it would be perfect if those features are included in your package at the end.

**Comment by dianjuar** [comment]
> I noticed that I missed some things to make it work for all Nx projects _including non-Angular libraries_. Actually the published directory is grabbed from the `dest` property in the project's `ng-package.json` file, we should fallback to the `outputPath` property in `workspace.json` definition file.

Yeah, this is really inconvenient because it will work perfectly for Angular and React Libs, but won't for node libraries 😞, we have to implement the way that you described to get the `outputPath`.
But I think that is not difficult to implement, I will give it a try this weekend.

**Comment by dianjuar** [comment]
> Also we need to ensure that this builder is idempotent to suit our needs: automated release cycle for a monorepo. We will iterate and create other PRs, but because we want this builder for our [@jscutlery/semver](https://github.com/jscutlery/semver) plugin, I will re-fork this project under our organization and point to our git while these PRs are pending review, it will be easier for us.
> 
> However it would be perfect if those features are included in your package at the end.

I will be honoured if you do so 😃

**Comment by dianjuar** [comment]
> @edbzn
> 
> > I noticed that I missed some things to make it work for all Nx projects _including non-Angular libraries_. Actually the published directory is grabbed from the `dest` property in the project's `ng-package.json` file, we should fallback to the `outputPath` property in `workspace.json` definition file.
> 
> Yeah, this is really inconvenient because it will work perfectly for Angular and React Libs, but won't for node libraries 😞, we have to implement the way that you described to get the `outputPath`. I think that is not difficult to implement though, I will give it a try this weekend 😄

@edbzn I could code the solution to your problem, I was going to merge your PR to add the solution but I discover that you accidentally added some other commits that don't belong to this PR. Could you please remove those commits to be able to merge your PR.

I can release the new version tomorrow night (UTC -8:00) so you can enjoy it 🔥 

image

**Comment by dianjuar** [comment]
@edbzn One more thing, I was thinking about the documentation of this feature but I'm pretty sure how to express it 🤔 
It's something about with this builder you can publish any library on an Nx workspace because it's a plugin now

Seems that you are more into Nx plugins than I, any ideas?

**Review by dianjuar — APPROVED** [review]
Thank you so much for your contribution 😄 🚀

**Comment by edbzn** [comment]
Hi @dianjuar, do you still want that I remove the last commits? I didn't finished yet at tested it very well. Concerning the documentation we coud just tell that Nx workspace is supported as well as traditional ng-cli projects. Now it relies on workspace/project definition.

**Comment by dianjuar** [comment]
> Hi @dianjuar, do you still want that I remove the last commits? I didn't finished and tested it very well yet. Concerning the documentation we coud just tell that Nx workspace is supported as well as traditional ng-cli projects. Now it relies on workspace/project definition.

I could make it work for Angular libs, React Libs and, Ts Libs. That is pretty much all the libs that you can create on an Nx Workspace

**Comment by edbzn** [comment]
Ok, sounds good! :blush:

**Comment by dianjuar** [comment]
@edbzn I deeply appreciate the work that you did here, it opened the possibility to extend this library to frontiers that I couldn't imagine.

Thank you so much for the work done here 👍

**Pull request merged**